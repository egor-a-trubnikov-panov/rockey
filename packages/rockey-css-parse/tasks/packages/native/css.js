import { argv } from 'yargs';
import chalk from 'chalk';

import getNativeCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import css from 'css';

const size = argv.size || 1000;
const CSSString = getNativeCss(size);

console.log('');
console.log(
  `Start ${chalk.green('css')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

timer(() => {
  const parsed = css.parse(CSSString);
  css.stringify(parsed);
}).then(time => {
  output('css', size, time);
});
