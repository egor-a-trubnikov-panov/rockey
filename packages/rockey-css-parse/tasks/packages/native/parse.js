import { argv } from 'yargs';
import chalk from 'chalk';

import getNativeCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import createParser from '../../../lib/parse';
import { stringifyRules } from '../../../lib/stringify';

const size = argv.size || 1000;
const css = getNativeCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

const parse = createParser();

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
