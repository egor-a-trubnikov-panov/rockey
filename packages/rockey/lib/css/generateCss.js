import mergeCss from './utils/mergeCss';

import {
  getClassName,
  getRootClassNameMap,
  generateAnimationName,
} from './getClassName';

const getSelector = (displayName, parent, mixin) => {
  if (
    displayName[0] !== '[' &&
    displayName[0] !== ':' &&
    displayName[0] !== '.' &&
    displayName[0] !== '#' &&
    displayName[0] === displayName[0].toUpperCase()
  ) {
    return `${mixin || ''}${getClassName(displayName, parent)}`;
  } else {
    return `${mixin || ''}${parent ? `${parent} ${displayName}` : `${displayName}`}`;
  }
};

const isMedia = key => key.indexOf('@media') === 0;
const isKeyFrames = key => key.indexOf('@keyframes') === 0;
const isNot = key => key.indexOf(':not') === 0;

// TODO: remove code duplic.
function process(styles, context) {
  let processed = styles;

  if (styles['animation-name']) {
    processed = { ...styles };

    let animationName = styles['animation-name'];
    if (!context.animations[animationName]) {
      if (animationName !== 'none' && animationName.indexOf('none ') !== 0) {
        console.warn(
          `(generate error) "${animationName}" animation was requested but had not been defined`
        );

        delete processed['animation-name'];
      }
    } else {
      processed['animation-name'] = context.animations[animationName];
    }
  } else if (styles['animation']) {
    processed = { ...styles };

    let animationName = styles['animation'].split(' ')[0];

    if (!context.animations[animationName]) {
      if (animationName !== 'none') {
        console.warn(
          `(generate error) "${animationName}" animation was requested but had not been defined`
        );

        delete processed.animation;
      }
    } else {
      processed['animation'] = styles['animation'].replace(
        animationName,
        context.animations[animationName]
      );
    }
  }

  return processed;
}

function processModificators(
  tree,
  {
    parent,
    context,
  }
) {
  const css = {};

  for (const modificatorKey of Object.keys(tree.modificators)) {
    const modificator = tree.modificators[modificatorKey];

    if (isMedia(modificatorKey)) {
      css[modificatorKey] = generateCss(modificator, {
        parent,
      });

      if (Object.keys(modificator.styles).length) {
        css[modificatorKey][parent] = modificator.styles;
      }
    } else if (isKeyFrames(modificatorKey)) {
      let animationName = modificatorKey.split(' ')[1];

      let uniqAnimationName = generateAnimationName(animationName);
      context.animations[animationName] = uniqAnimationName;

      const keyframes = Object.keys(modificator.components).reduce(
        (steps, step) => {
          steps[step] = modificator.components[step].styles;
          return steps;
        },
        {}
      );

      css[modificatorKey.replace(animationName, uniqAnimationName)] = keyframes;
    } else {
      let updatedModificatorKey = modificatorKey;

      if (isNot(modificatorKey)) {
        const matches = modificatorKey.match(/\(([^\)]+)\)(.*)/);
        updatedModificatorKey = `:not(${getSelector(matches[1])})${matches[2]}`;
      }

      // merge?
      if (Object.keys(modificator.styles).length) {
        css[`${parent}${updatedModificatorKey}`] = process(
          modificator.styles,
          context
        );
      }

      Object.assign(
        css,
        generateCss(modificator, {
          parent: `${parent}${updatedModificatorKey}`,
        })
      );
    }
  }

  return css;
}

function generateCss(
  tree,
  {
    parent,
    mixin,
    context,
  }
) {
  const css = {};

  if (parent && Object.keys(tree.modificators)) {
    // render parent modificators
    mergeCss(
      css,
      processModificators(tree, {
        context,
        parent,
      })
    );
  }

  for (const className of Object.keys(tree.components)) {
    const component = tree.components[className];
    const selector = getSelector(className, parent, mixin);

    mergeCss(
      css,
      generateCss(component, {
        parent: selector,
        context,
      })
    );

    if (Object.keys(component.styles).length) {
      css[selector] = process(component.styles, context);
    }
  }

  return css;
}

export default (
  tree,
  {
    isRoot = true,
    parent = null,
    mixin = null,
    context = {
      animations: {},
    },
  } = {}
) => {
  if (Object.keys(tree.styles).length) {
    throw new Error('(generate css) root tree should not contain styles');
  }

  const rootCss = processModificators(tree, {
    context,
    parent,
  });

  const classNameMap = getRootClassNameMap(tree, {
    isRoot,
  });

  return {
    css: mergeCss(
      rootCss,
      generateCss(tree, {
        parent,
        mixin,
        context,
      })
    ),
    context,
    classNameMap: classNameMap,
  };
};
