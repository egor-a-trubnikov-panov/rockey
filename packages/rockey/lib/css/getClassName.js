import hash from '../utils/hash';

let comopnentsClassList = new Map();

// For correct styling - this Map should be empty when App is loaded.
// If not - styles are wrong
let notYetDefinedAtRootClassList = new Map();

const generateClassName = displayName => `${displayName}-${hash()}`;

export const clearCachedClassNames = () => {
  comopnentsClassList = new Map();
  notYetDefinedAtRootClassList = new Map();
};

export const getNotYetDefiendComponents = () => {
  const components = [];
  notYetDefinedAtRootClassList.forEach((className, displayName) =>
    components.push(displayName));

  return components;
};

export const getClassName = (displayName, parent, pre = '') => {
  let className = null;

  if (!comopnentsClassList.has(displayName)) {
    if (notYetDefinedAtRootClassList.has(displayName)) {
      className = notYetDefinedAtRootClassList.get(displayName);
    } else {
      className = generateClassName(displayName, false);
      notYetDefinedAtRootClassList.set(displayName, className);
    }
  } else {
    className = comopnentsClassList.get(displayName);
  }

  return parent ? `${parent} ${pre}.${className}` : `${pre}.${className}`;
};

export default getClassName;

export const getRootClassNameMap = (tree, { isRoot }) => {
  const classNameMap = {};

  // Root components should be unique
  // Two cases of Root Component className generation:
  // - at getRootClassNameMap function
  // - at getClassName if Comopnent is described as child at another root component
  for (const displayName of Object.keys(tree.components)) {
    if (!comopnentsClassList.has(displayName)) {
      let className = null;
      if (notYetDefinedAtRootClassList.has(displayName)) {
        className = notYetDefinedAtRootClassList.get(displayName);
        notYetDefinedAtRootClassList.delete(displayName);
      } else {
        className = generateClassName(displayName);
      }

      classNameMap[displayName] = className;
      comopnentsClassList.set(displayName, className);
    } else {
      if (isRoot) {
        console.warn(
          `(generate error) "${displayName}" was already defined as root component`
        );
      }

      classNameMap[displayName] = comopnentsClassList.get(displayName);
    }
  }

  return classNameMap;
};

// let mixinCounter = 0;
export const generateMixinClassName = mixinFunc => {
  const mixinName = mixinFunc.displayName || mixinFunc.name || 'anon';
  return `Mixin-${mixinName}-${hash()}`;
};

export const generateAnimationName = animationName =>
  `${animationName}-${hash()}`;
