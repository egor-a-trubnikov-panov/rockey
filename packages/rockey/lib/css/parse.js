// const validateRule = (rule, value) => {};
// const autoprefixRule = (rule, value) => {};

const cache = new Map();

const parseCss = styles => {
  const stylesParts = styles
    .trim()
    .split(';')
    .map(s => s ? s.split(':') : null)
    .filter(s => !!s);

  const key = stylesParts.toString();

  if (!key.trim()) {
    return {};
  }

  if (cache.has(key)) {
    return cache.get(key);
  }

  const stylesJson = stylesParts.reduce(
    (json, value) => {
      if (value) {
        const key = value[0].trim();
        if (json[key]) {
          // css fallback values
          let prev = Array.isArray(json[key]) ? json[key] : [json[key]];

          json[value[0].trim()] = prev.concat([
            value.slice(1).join(':').trim(),
          ]);
        } else {
          json[value[0].trim()] = value.slice(1).join(':').trim();
        }

        // validateRule(value[0], json[value[0]]);
        // json[value[0].trim()] = autoprefixRule(value.slice(1).join(':').trim());
      }
      return json;
    },
    {}
  );

  cache.set(key, stylesJson);

  return stylesJson;
};

// @replace-start
const isModificatorStartSymbol = symbol => {
  return symbol === '@' || symbol === ':' || symbol === '[';
};
// @replace-end

// @replace-start
const isModificatorEndSymbol = symbol => {
  return symbol === '{' || symbol === ';';
};
// @replace-end

// TODO: make this func beauty :)
const isStartsWithModificator = raw => {
  return 0 === raw.indexOf('@media') ||
    0 === raw.indexOf('@keyframes') ||
    0 === raw.indexOf('::after') ||
    0 === raw.indexOf('::before') ||
    0 === raw.indexOf('::first-letter') ||
    0 === raw.indexOf('::first-line') ||
    0 === raw.indexOf(':active') ||
    0 === raw.indexOf(':checked') ||
    0 === raw.indexOf(':disabled') ||
    0 === raw.indexOf(':empty') ||
    0 === raw.indexOf(':enabled') ||
    0 === raw.indexOf(':first-child') ||
    0 === raw.indexOf(':first-of-type') ||
    0 === raw.indexOf(':focus') ||
    0 === raw.indexOf(':hover') ||
    0 === raw.indexOf(':in-range') ||
    0 === raw.indexOf(':invalid') ||
    0 === raw.indexOf(':lang') ||
    0 === raw.indexOf(':last-child') ||
    0 === raw.indexOf(':last-of-type') ||
    0 === raw.indexOf(':link') ||
    0 === raw.indexOf(':not') ||
    0 === raw.indexOf(':nth-child') ||
    0 === raw.indexOf(':nth-last-child') ||
    0 === raw.indexOf(':nth-last-of-type') ||
    0 === raw.indexOf(':nth-of-type') ||
    0 === raw.indexOf(':only-of-type') ||
    0 === raw.indexOf(':only-child') ||
    0 === raw.indexOf(':optional') ||
    0 === raw.indexOf(':out-of-range') ||
    0 === raw.indexOf(':read-only') ||
    0 === raw.indexOf(':read-write') ||
    0 === raw.indexOf(':required') ||
    0 === raw.indexOf(':root') ||
    0 === raw.indexOf(':target') ||
    0 === raw.indexOf(':valid') ||
    0 === raw.indexOf(':visited');
};

const parse = (raw, parent) => {
  let openedBrackets = 0;
  let openedModificatorBrackets = 0;
  let styles = '';

  let component = null;
  let combinedComponents = [];
  let components = {};

  let possibleModificator = 0;
  let modificators = {};
  let modificator = '';

  let current = '';
  let currentBackup = '';

  // @replace-start
  const restore = () => {
    current = currentBackup + current;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const clearAndRestore = () => {
    current = currentBackup;
    currentBackup = '';
  };
  // @replace-end

  // @replace-start
  const backup = () => {
    currentBackup = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const possibleModificatorStart = () => {
    possibleModificator = true;
  };
  // @replace-end

  // @replace-start
  const isPossibleModificator = () => {
    return possibleModificator;
  };
  // @replace-end

  // @replace-start
  const possibleModificatorEnd = () => {
    possibleModificator = false;
  };
  // @replace-end

  // @replace-start
  const isModificator = () => {
    return isStartsWithModificator(current.trim());
  };
  // @replace-end

  // @replace-start
  const openBracket = () => {
    openedBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeBracket = () => {
    openedBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllBracketsClosed = () => {
    return openedBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const openModificatorBracket = () => {
    openedModificatorBrackets++;
  };
  // @replace-end

  // @replace-start
  const closeModificatorBracket = () => {
    openedModificatorBrackets--;
  };
  // @replace-end

  // @replace-start
  const isAllModificatorBracketsClosed = () => {
    return openedModificatorBrackets === 0;
  };
  // @replace-end

  // @replace-start
  const startModificator = () => {
    modificator = current;
    current = '';
  };
  // @replace-end

  // @replace-start
  const saveModificator = () => {
    modificator = modificator.trim();

    // @remove-start
    if (modificators[modificator]) {
      throw new Error(
        `(parse error) "${modificator}" duplicated definiton at one block`
      );
    }
    // @remove-end

    modificators[modificator] = parse(current, {
      parentType: 'modificator',
      name: modificator,
    });

    modificator = '';
  };
  // @replace-end

  // @replace-start
  const isInsideModificator = () => {
    return !!modificator;
  };
  // @replace-end

  // @replace-start
  const isRootComponentBlockStarts = symbol => {
    return symbol === '{' && isAllBracketsClosed();
  };
  // @replace-end

  // @replace-start
  const startComponent = () => {
    let parts = current.trim().split(' ');
    if (parts.length === 1) {
      component = parts[0];
    } else {
      combinedComponents = [];

      let i = 0;
      for (i = parts.length - 1; i >= 0; i--) {
        const part = parts[i];
        const charCode = part.charCodeAt(0);

        /* All chars starts with uppercase or xxx% for animations  */
        if (
          (charCode >= 65 && charCode <= 90) || part[part.length - 1] === '%'
        ) {
          combinedComponents.push(part.replace(',', '').trim());
        } else if (part[0] === '~' || part[0] === '+') {
          let size = combinedComponents.length;
          combinedComponents[size - 1] = part[0] + combinedComponents[size - 1];
        } else {
          break;
        }
      }

      if (combinedComponents.length) {
        component = combinedComponents[0];
        combinedComponents = combinedComponents.slice(1);
        parts = parts.slice(0, i + 1);
      } else {
        component = parts[parts.length - 1];
        parts = parts.slice(0, -1);
      }

      styles += parts.join(' ');
    }

    current = '';
  };
  // @replace-end

  // @replace-start
  const isInsideComponent = () => {
    return !!component;
  };
  // @replace-end

  // @replace-start
  const save = symbol => {
    current += symbol;
  };
  // @replace-end

  // @replace-start
  const saveComponent = () => {
    // @remove-start
    if (components[component]) {
      throw new Error(
        `(parse error) "${component}" duplicated definiton at one block`
      );
    }
    // @remove-end
    components[component] = parse(current, {
      parentType: 'component',
      name: component,
      combinedComponents,
    });

    combinedComponents = [];

    component = null;
    current = '';
  };
  // @replace-end

  // @replace-start
  const updateStylesTail = () => {
    if (current) {
      styles += current;
    }
  };
  // @replace-end

  // ---
  const mixins = [];
  let possibleMixin = false;
  let mixin = '';

  // @replace-start
  const isMixinStartSymbol = symbol => {
    return symbol === '_';
  };
  // @replace-end

  // @replace-start
  const isMixinEndSymbol = symbol => {
    return symbol === ' ';
  };
  // @replace-end

  // @replace-start
  const possibleMixinStart = () => {
    possibleMixin = true;
  };
  // @replace-end

  // @replace-start
  const possibleMixinEnd = () => {
    possibleMixin = false;
  };
  // @replace-end

  // @replace-start
  const isPossibleMixin = () => {
    return possibleMixin;
  };
  // @replace-end

  // @replace-start
  const saveMixinString = symbol => {
    mixin += symbol;
  };
  // @replace-end

  // @replace-start
  const clearMixin = () => {
    mixin = '';
  };
  // @replace-end

  // @replace-start
  const removeMixinFromStyles = () => {
    current = current.replace(mixin.trim(), '');
  };
  // @replace-end

  // @replace-start
  const saveMixin = () => {
    mixins.push(mixin.trim());
  };
  // @replace-end

  // @remove-start
  const isInValidMixinPosition = () => {
    return !parent ||
      (currentBackup.trim().length &&
        !new RegExp(`(;|{)\\s+?${mixin.trim()}`).test(`${current}`.trim()));
  };
  // @remove-end

  // ----

  for (let symbol of raw) {
    if (!isInsideComponent() && !isInsideModificator()) {
      if (isMixinStartSymbol(symbol)) {
        possibleMixinStart();
      }

      if (isPossibleMixin()) {
        saveMixinString(symbol);
        if (isMixinEndSymbol(symbol)) {
          possibleMixinEnd();
          // @remove-start
          if (isInValidMixinPosition()) {
            if (parent) {
              const nearestCode = `${currentBackup}${current}`
                .replace(/\s+/g, ' ')
                // .replace(mixin.trim(), '{{ MIXIN }}' + mixin)
                .trim();

              console.warn(
                'seems wrong mixin position',
                `${parent.name} - ${nearestCode}`
              );
            } else {
              console.warn('mixin should not be at root');
            }
          }
          // @remove-end

          saveMixin();
          removeMixinFromStyles();
          clearMixin();
        }
      }
    }

    if (!isInsideComponent() && !isInsideModificator()) {
      if (!isPossibleModificator() && isModificatorStartSymbol(symbol)) {
        possibleModificatorStart();
        backup();
      }

      if (isPossibleModificator() && isModificatorEndSymbol(symbol)) {
        if (isModificator()) {
          startModificator();
        } else {
          restore();
        }

        possibleModificatorEnd();
      }
    }

    if (
      !isPossibleModificator() &&
      !isInsideModificator() &&
      isRootComponentBlockStarts(symbol)
    ) {
      openBracket();
      startComponent();
      // continue;
    } else if (symbol === '{') {
      if (isInsideComponent()) {
        openBracket();

        if (openedBrackets !== 1) {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        openModificatorBracket();

        if (openedModificatorBrackets !== 1) {
          save(symbol);
        }
      }
    } else if (symbol === '}') {
      if (isInsideComponent()) {
        closeBracket();
      } else if (isInsideModificator()) {
        closeModificatorBracket();
      }

      if (isInsideComponent()) {
        if (isAllBracketsClosed()) {
          saveComponent();
        } else {
          save(symbol);
        }
      } else if (isInsideModificator()) {
        if (isAllModificatorBracketsClosed()) {
          saveModificator();
          clearAndRestore();
        } else {
          save(symbol);
        }
      }
    } else {
      save(symbol);
    }
  }

  restore();
  updateStylesTail();

  return {
    mixins,
    components,
    combinedComponents: (parent && parent.combinedComponents) || [],
    styles: parseCss(styles),
    modificators,
  };
};

export default inline => {
  return parse(inline.replace(/\r|\n/g, '').replace(/\s+/g, ' '));
};
