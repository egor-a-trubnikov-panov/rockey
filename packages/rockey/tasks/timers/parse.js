import { argv } from 'yargs';
import chalk from 'chalk';

import getCss from './utils/getCss';
import timer from './utils/timer';
import output from './utils/output';

import parse from '../../lib/css/parse';
import { stringifyRules } from '../../lib/styleSheets/utils/stringify';

const size = argv.size || 10;
const css = getCss(size);

console.log('');
console.log(
  `Start ${chalk.green('parse')} measuring with ${chalk.cyan(size)} class length`
);
console.log('');

timer(() => {
  stringifyRules(parse(css).precss);
}).then(time => {
  output('parse', size, time);
});
