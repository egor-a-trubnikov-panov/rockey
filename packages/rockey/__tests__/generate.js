import { styles } from './parse';

import parse from '../lib/css/parse';
import generateCss from '../lib/css/generateCss';
import * as getClassNameModule from '../lib/css/getClassName';

const getClassName = getClassNameModule.default;

console.warn = jest.fn();

const classname = className => `.${className}`;

describe('generate css', () => {
  it('empty', () => {
    const generated = generateCss(parse(styles.empty));
    expect(generated.css).toEqual({});
  });

  it('onlyStyles', () => {
    expect(() => generateCss(parse(styles.onlyStyles))).toThrow(
      '(generate css) root tree should not contain styles'
    );
  });

  it('withComponents', () => {
    const generated = generateCss(parse(styles.withComponents));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
    });
  });

  it('withNestedComponents', () => {
    const generated = generateCss(parse(styles.withNestedComponents));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}`]: {
        color: 'blue',
      },
    });
  });

  it('withModificators', () => {
    const generated = generateCss(parse(styles.withModificators));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)}:hover`]: {
        color: 'purple',
      },
      [`${classname(classNameMap.Button)}::before`]: {
        width: '16px',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}`]: {
        color: 'blue',
      },
    });
  });

  it('withComponentsInsideModificators', () => {
    const generated = generateCss(
      parse(styles.withComponentsInsideModificators)
    );
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)}:hover`]: {
        color: 'purple',
      },
      [`${classname(classNameMap.Button)}:hover ${getClassName('Icon')}`]: {
        color: 'blue',
      },
    });
  });

  it('withModificatorsInsideComponents', () => {
    const generated = generateCss(
      parse(styles.withModificatorsInsideComponents)
    );
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}`]: {
        color: 'blue',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}:hover`]: {
        color: 'purple',
      },
    });
  });

  it('withModificatorsInsideComponents', () => {
    const generated = generateCss(
      parse(styles.withModificatorsInsideComponents)
    );
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}`]: {
        color: 'blue',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}:hover`]: {
        color: 'purple',
      },
    });
  });

  it('withMedia', () => {
    const generated = generateCss(parse(styles.withMedia));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}`]: {
        color: 'blue',
      },
      [`${classname(classNameMap.Button)} ${getClassName('Icon')}:hover`]: {
        color: 'purple',
      },
      '@media (max-width: 699px)': {
        [`${classname(classNameMap.Button)} ${getClassName('Icon')}`]: {
          background: 'green',
        },
        [`${classname(classNameMap.Button)} ${getClassName('Icon')}:hover`]: {
          color: 'purple',
        },
      },
      '@media screen and (max-width: 699px) and (min-width: 520px)': {
        [classname(classNameMap.Button)]: {
          background: 'yellow',
        },
      },
    });
  });

  it('multipleMediaAtOneLevel', () => {
    const generated = generateCss(parse(styles.multipleMediaAtOneLevel));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      '@media (max-width: 699px)': {
        [classname(classNameMap.Button)]: {
          background: 'green',
        },
        [classname(classNameMap.Button2)]: {
          background: 'blue',
        },
        [`${classname(classNameMap.Button)}:hover`]: {
          color: 'green',
        },
        [`${classname(classNameMap.Button2)}:hover`]: {
          color: 'blue',
        },
      },
    });
  });

  it('withFallbackValues', () => {
    const generated = generateCss(parse(styles.withFallbackValues));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: ['rgba(0, 255, 0, 0.3)', '#00ff00'],
      },
    });
  });

  it('withKeyframes', () => {
    let animationCounter = 0;

    const originalGenerateAnimationName = getClassNameModule.generateAnimationName;
    getClassNameModule.generateAnimationName = animationName => {
      return `${animationName}-a${++animationCounter}`;
    };

    const generated = generateCss(parse(styles.withKeyframes));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        animation: 'example-first-a1 1s infinity',
      },

      [classname(classNameMap.Button2)]: {
        'animation-duraiton': '1s',
        'animation-iteration-count': 'infinity',
        'animation-name': 'example-second-a2',
      },

      '@keyframes example-first-a1': {
        '0%': { color: 'red' },
        '50%': { color: 'yellow' },
        '100%': { color: 'red' },
      },

      '@keyframes example-second-a2': {
        '0%': { color: 'purple' },
        '100%': { color: 'black' },
      },
    });

    getClassNameModule.generateAnimationName = originalGenerateAnimationName;
  });

  it('withTagSelectors', () => {
    const generated = generateCss(parse(styles.withTagSelectors));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)} span.text`]: {
        color: 'black',
      },
      [`${classname(classNameMap.Button)} div#hardcode`]: {
        color: 'white',
      },
    });
  });

  it('withMostModificators', () => {
    const generated = generateCss(parse(styles.withMostModificators));
    const classNameMap = generated.classNameMap;

    expect(generated.css).toEqual({
      [classname(classNameMap.Button)]: {
        color: 'red',
      },
      [`${classname(classNameMap.Button)}::first-line`]: {
        'background-color': 'yellow',
      },
      [`${classname(classNameMap.Button)}:disabled`]: {
        color: 'green',
      },
      [`${classname(classNameMap.Button)}:hover`]: {
        color: 'blue',
      },
      [`${classname(classNameMap.Button)}::after`]: {
        color: 'purple',
      },
      [`${classname(classNameMap.Button)}:nth-child(2)`]: {
        color: 'white',
      },
      [`${classname(classNameMap.Button)}:not(.hardcodeClass)`]: {
        color: 'yellow',
      },
      [`${classname(classNameMap.Button)}:not(${getClassName('Icon')})`]: {
        color: 'orange',
      },
    });
  });
});

test('generate: customStyles', () => {
  const generated = generateCss(
    parse(
      `
    Button {
      font-size: 14px;
      line-height: 36px;
      box-sizing: border-box;
      text-transform: uppercase;
      border-radius: 3px;
      user-select: none;
      border: 0;
      font-weight: 500;
      letter-spacing: .010em;
      outline: 0;
      cursor: pointer;
      overflow: hidden;
      display: inline-block;
      white-space: nowrap;
      padding: 0 6px;
      margin: 6px 8px;
      min-height: 36px;
      min-width: 88px;
      text-align: center;
      transition:
        box-shadow .4s cubic-bezier(.25,.8,.25,1),
        background-color .4s cubic-bezier(.25,.8,.25,1);

      background: transparent;

      :not(:disabled):hover {
        background-color: transparent
      }

      :focus {
        outline: 0;
      }

      :disabled {
        color: #ccc;
        background: transparent;
      }
    }
  `
    )
  );

  expect(generated.css).toEqual({
    [classname(generated.classNameMap.Button)]: {
      background: 'transparent',
      border: '0',
      'border-radius': '3px',
      'box-sizing': 'border-box',
      cursor: 'pointer',
      display: 'inline-block',
      'font-size': '14px',
      'font-weight': '500',
      'letter-spacing': '.010em',
      'line-height': '36px',
      margin: '6px 8px',
      'min-height': '36px',
      'min-width': '88px',
      outline: '0',
      overflow: 'hidden',
      padding: '0 6px',
      'text-align': 'center',
      'text-transform': 'uppercase',
      transition: 'box-shadow .4s cubic-bezier(.25,.8,.25,1),        background-color .4s cubic-bezier(.25,.8,.25,1)',
      'user-select': 'none',
      'white-space': 'nowrap',
    },
    [classname(generated.classNameMap.Button) + ':disabled']: {
      background: 'transparent',
      color: '#ccc',
    },
    [classname(generated.classNameMap.Button) + ':focus']: {
      outline: '0',
    },
    [classname(generated.classNameMap.Button) + ':not(:disabled):hover']: {
      'background-color': 'transparent',
    },
  });
});
