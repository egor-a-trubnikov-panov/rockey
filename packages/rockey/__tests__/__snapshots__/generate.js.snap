// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse: customStyles 1`] = `
Object {
  "combinedComponents": Array [],
  "components": Object {
    "Button": Object {
      "combinedComponents": Array [],
      "components": Object {
        "Icon": Object {
          "combinedComponents": Array [],
          "components": Object {},
          "mixins": Array [],
          "modificators": Object {},
          "styles": Object {
            "color": "#fff",
          },
        },
      },
      "mixins": Array [],
      "modificators": Object {},
      "styles": Object {
        "box-shadow": "0 2px 5px rgba(0, 0, 0, .26)",
        "margin": "5px 6px",
      },
    },
  },
  "mixins": Array [],
  "modificators": Object {},
  "styles": Object {},
}
`;
